lib/connectors/memory.js

```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
   });
 };
 
+Memory.prototype.upsertWithWhere = function(model,data,options,callback){
+
+var self = this;
+  try {
+
+    var response;
+
+    var element = Object.keys(where)[0];
+
+    var filter = {
+      "where": where
+    };
+
+
+    var response = null;
+    if (Object.keys(data).length === 0)
+      response = false;
+    else
+      response = true;
+
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
+    }
+
+    if (response == true) {
+
+      self.all(filter, function (err, models) {
+
+        if (err) {
+
+          callback(err);
```

Comment:
Please wrap `callback` into `process.nextTick()` so that callback only happens after the function is finished.